Use string representation of response data

viktordick opened the pull request:
Response data that has no `__bytes__` method but a `__str__` method,
like lists, dicts and similar, can now be handled by the publisher as it
was handled in Zope 2, by returning the string representation.

Fixes https://github.com/zopefoundation/Zope/issues/1006

viktordick commented:
Tested this on the system where I encountered the problem described in #1006 and it seems to work. I am in the process of executing the tests.

viktordick commented:
One test fails on my system, but it also fails on `master`.

@d-maurer If you think this might be a solution, I could add a test for the situation I described in #1006 and mark this ready for review.

d-maurer commented on CHANGES.rst:
```diff
@@ -13,6 +13,9 @@ https://github.com/zopefoundation/Zope/blob/4.x/CHANGES.rst
 
 - Update to newest compatible versions of dependencies.
 
+- Use string representation of response data that has no ``__bytes__``
```
Suggestion: "Use intermediate ``str`` representation for complex response data (i.e. not `str` or bytes like)

d-maurer reviewed (commented):
Could you also improve the docstring? It does not reflect the implementation (`asHTML` is used without preceding type check) and still uses Python 2 terms.

icemac commented:
I think this is a good approach. After the comments of @d-maurer are fixed, I'd like to approve this PR.

viktordick commented:
@icemac thanks. I adjusted the docstring and added a test. While writing the docstring, I was wondering if there are cases where `body` has no `__bytes__` method, but we *still* want to use `bytes(body)` instead of `self._encode_unicode(str(body))`. If there is no such situation, I guess a more transparent order of operations would be something like
```
if not isinstance(body, bytes) and hasattr(body, '__bytes__'):
    try:
        body = bytes(body)
    except (TypeError, UnicodeError):
        pass
if not isinstance(body, bytes):
    body = self._encode_unicode(str(body))
```
to replace the complete if-chain.
I am aware that the current state of the PR is less invasive that this, I just wanted to mention my thoughts after reading the code.

d-maurer commented:
Viktor Dick wrote at 2022-1-16 23:41 -0800:
***@***.*** thanks. I adjusted the docstring and added a test. While writing the docstring, I was wondering if there are cases where `body` has no `__bytes__` method, but we *still* want to use `bytes(body)` instead of `self._encode_unicode(str(body))`.

There are.
An example is an instance of `io.BytesIO`: it has no `__bytes__` method
but logically contains a sequence of bytes (returned by `bytes(bio)`).
It is a natural choice when the application wants to return
a bytes response constructed incrementally.
```pycon:
>>> from io import BytesIO
>>> s=BytesIO()
>>> s.__bytes__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: '_io.BytesIO' object has no attribute '__bytes__'
>>> bytes(s)
b''
```
Please verify that your PR can handle such objects correctly.


I had a look (in Python 3.9) at `bytes_new`
(essentially the implementation of `bytes(obj)`). It (essentially)
checks: "has `__bytes__`", "unicode",
"bytes like", "buffer like", "list", "tuple".

viktordick commented:
Well, I just tried the following:
```
>>> from io import BytesIO
>>> d = BytesIO(b'test')
>>> bytes(d)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'bytes' object cannot be interpreted as an integer
>>> d.getvalue()
b'test'
```
so `bytes(d)` does not in fact return the content of the `BytesIO` object. Am I missing something?

viktordick commented:
That is, if I want to use `BytesIO` to incrementally construct the response data, I need to
```
return result.getbuffer()
```
both before and after this PR.

viktordick commented:
OK sorry for the many messages.
In fact, `BytesIO.getbuffer` returns a `memoryview`, which has no `__bytes__` but allows for `bytes(...)` - not sure how. So yes, this is probably a regression of this approach.
I will have to think about it.

d-maurer commented:
Viktor Dick wrote at 2022-1-17 00:37 -0800:
>Well, I just tried the following:
>```
>>>> from io import BytesIO
>>>> d = BytesIO(b'test')
>>>> bytes(d)
>Traceback (most recent call last):
>  File "<stdin>", line 1, in <module>
>TypeError: 'bytes' object cannot be interpreted as an integer
>>>> d.getvalue()
>b'test'
>```
>so `bytes(d)` does not in fact return the content of the `BytesIO` object. Am I missing something?

Apparently, I have been wrong: Python does not support `bytes(BytesIO)`,
at least not unless the `BytesIO` is empty.

I expected `BytesIO` to implement the "buffer protocol".
To my surprise, it does not do that.
Therefore, `bytes(BytesIO)` does not take the "buffer like" case
but the "iterator like" case. In this case, the iterator is expected
to generate a sequence of integers but `BytesIO` generates a sequence
of bytes which leads to the `TypeError` above.


Nevertheless, there are objects without `__bytes__` for which `bytes(obj)`
succeeds. New example `bytearray`:
```pycon:
>>> b=bytearray(b'test')
>>> b.__bytes__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'bytearray' object has no attribute '__bytes__'
>>> bytes(b)
b'test'


Further candidates are all objects supporting the "buffer protocol".

viktordick commented:
> Nevertheless, there are objects without `__bytes__` for which `bytes(obj)` succeeds. New example `bytearray`:

Well this becomes difficult then. Because `bytes([1, 2, 3])` also succeeds, but `bytes([1, 2, 500])` does not. But I would not want a different handling of lists depending on if the values happen to be integers in the range `[0,255]`.

But I guess we will have to do one of the sacrifices.
Solution 1: We simply try `bytes(body)` and fall back to `self._encode_unicode(str(body))` in case of errors, which will give a significantly different handling of `[1, 2, 3]` than `[1, 2, 500]`.
Solution 2: We use a specific list of types that are converted to `bytes` directly and use the string representation otherwise. The list would contain `bytes`, `memoryview`, `bytearray` and possibly others (I tried to check for how to know if a type implements the buffer protocol, but could not find anything clear).

Side note: I would in fact also be able to live with solution 1. My concrete problem is that the transaction rolls back due the error. The actual result is not really used, only stored for debugging purposes, so I could live with such inconsistencies. If the result of a method that is called by URL is actually to be used and it is not a simple str or bytes, the method should make sure to pack the result accordingly and not depend on the specifics of the `ZPublisher` (i.e., it should probably pack the list into JSON).
But I think solution 2 is more elegant.

viktordick commented:
@d-maurer Thanks for the suggestion, I implemented it, extending the first point ("If it is a `bytes` instance") to also include `bytearray` and `memoryview`.

`setBody` will still raise a `ValueError` if for example a sequence is passed with non-integers or integers outside the range of [0, 255] after explicitly setting the content type to expect binary output, but I guess this is expected and actually a good thing.
